File: src/types/external/robotty.ts
// Mirrors Rust dto::external::robotty — snake_case wire format (matches the
// EventSub `channel.chat.message` shape so renderers can consume both).

export type EmoteRef = {
  id: string;
};

export type MentionRef = {
  user_login: string;
};

export type Fragment =
  | { type: "text"; text: string }
  | { type: "emote"; text: string; emote: EmoteRef }
  | { type: "mention"; text: string; mention: MentionRef };

export type MessageBody = {
  text: string;
  fragments: Fragment[];
};

export type Badge = {
  set_id: string;
  id: string;
  info: string;
};

export type Reply = {
  parent_message_id: string;
  parent_message_body: string;
  parent_user_name: string;
  parent_user_login: string;
};

export type RecentMessage = {
  broadcaster_user_id: string;
  message_id: string;
  chatter_user_id: string;
  chatter_user_login: string;
  chatter_user_name: string;
  color: string;
  message: MessageBody;
  message_type: string;
  badges: Badge[];
  reply: Reply | null;
  channel_points_custom_reward_id: string | null;
  timestamp_ms: number;
  deleted: boolean;
};

File: src-tauri/src/dto/external/robotty.rs
use serde::Serialize;

#[derive(Serialize)]
pub struct RecentMessage {
    pub broadcaster_user_id: String,
    pub message_id: String,
    pub chatter_user_id: String,
    pub chatter_user_login: String,
    pub chatter_user_name: String,
    pub color: String,
    pub message: MessageBody,
    pub message_type: String,
    pub badges: Vec<Badge>,
    pub reply: Option<Reply>,
    pub channel_points_custom_reward_id: Option<String>,
    pub timestamp_ms: i64,
    pub deleted: bool,
}

#[derive(Serialize)]
pub struct MessageBody {
    pub text: String,
    pub fragments: Vec<Fragment>,
}

#[derive(Serialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum Fragment {
    Text { text: String },
    Emote { text: String, emote: EmoteRef },
    Mention { text: String, mention: MentionRef },
}

#[derive(Serialize)]
pub struct EmoteRef {
    pub id: String,
}

#[derive(Serialize)]
pub struct MentionRef {
    pub user_login: String,
}

#[derive(Serialize)]
pub struct Badge {
    pub set_id: String,
    pub id: String,
    pub info: String,
}

#[derive(Serialize)]
pub struct Reply {
    pub parent_message_id: String,
    pub parent_message_body: String,
    pub parent_user_name: String,
    pub parent_user_login: String,
}
